go/mysql/conn_test.go

```diff
@@ -500,3 +504,145 @@ func (m mockAddress) String() string {
 }
 
 var _ net.Addr = (*mockAddress)(nil)
+
+var letters = []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ")
+
+func randSeq(n int) string {
+	b := make([]rune, n)
+	for i := range b {
+		b[i] = letters[rand.Intn(len(letters))]
+	}
+	return string(b)
+}
+
+func TestPrepareAndExecute(t *testing.T) {
+	// this test starts a lot of clients that all send prepared statement parameter values
+	// and check that the handler received the correct input
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	for i := 0; i < 1000; i++ {
+		startGoRoutine(ctx, t, randSeq(i))
+	}
+
+	ctx2, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	for {
+		select {
+		case <-ctx2.Done():
+			return
+		default:
+			if t.Failed() {
+				return
+			}
+		}
+	}
+}
+
+func startGoRoutine(ctx context.Context, t *testing.T, s string) {
+	go func(longData string) {
+		listener, sConn, cConn := createSocketPair(t)
+		defer func() {
+			listener.Close()
+			sConn.Close()
+			cConn.Close()
+		}()
+
+		sql := "SELECT * FROM test WHERE id = ?"
+		mockData := preparePacket(t, sql)
+
+		err := cConn.writePacket(mockData)
+		require.NoError(t, err)
+
+		handler := &testRun{
+			t:              t,
+			expParamCounts: 1,
+			expQuery:       sql,
+			expStmtID:      1,
+		}
+
+		ok := sConn.handleNextCommand(handler)
+		require.True(t, ok, "oh noes")
+
+		resp, err := cConn.ReadPacket()
+		require.NoError(t, err)
+		require.EqualValues(t, 0, resp[0])
+
+		for count := 0; ; count++ {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				if t.Failed() {
+					return
+				}
```

Comment:
this check is not interesting. as soon as this test fails. The context is cancelled and all the goroutines will start returning.